Wishcard edit/delete functionality

Jacobjeevan opened the pull request:
#357  

Adding edit/delete functionality to my wishes page; allowing agency users to edit their respective wishcards. Inactive wishcards are limited to delete only. Rest (draft/active) have both functionality.

- Added edit/delete routes
   - Added edit form page to frontend
- Added respective tests
- TODO: Complete styling for wishcard edit form
- TODO: Complete styling for buttons on my wishes page
- TODO: add confirmation dialog for delete functionality
- TODO (potentially): add ability to edit wishcard image.

Ability to edit wishcard image is the reason why it was delayed, couldn't get it work properly (request body had no elements; I tried setting different headers, but no luck). So removed that for now. Could revisit for future if necessary.

I am delaying the TODOs for now as this feature has already been going for a while, don't want to delay it any further. I will open up another branch to complete the todo's.

Enubia commented on client/public/clientJS/agencyWishCards.js:
```diff
@@ -0,0 +1,29 @@
+window.onload = function () {
```
use $(document).ready from jquery for this

Jacobjeevan replied:
Does this make any difference lol? (If following convention, I pretty much copied that from another js file, we aren't following any convention there :P).

I will change it though ;)

Enubia commented on client/public/clientJS/agencyWishCards.js:
```diff
@@ -0,0 +1,29 @@
+window.onload = function () {
+  let deleteBtns = document.getElementsByClassName('wishCardDeleteBtn');
```
$('wishCardDeleteBtn')

Enubia commented on client/public/clientJS/wishcardEdit.js:
```diff
@@ -0,0 +1,81 @@
+window.onload = function () {
```
use $(document).ready from jquery for this

Enubia commented on client/public/clientJS/wishcardEdit.js:
```diff
@@ -0,0 +1,81 @@
+window.onload = function () {
+  let agencyAddressCheckBox = document.getElementById('agencyAddressCheckBox');
+  agencyAddressCheckBox.addEventListener('click', handleCheckBoxClick);
+
```
make use of jquery

stacysealky replied:
^ lolol

Enubia commented on client/public/clientJS/wishcardEdit.js:
```diff
@@ -0,0 +1,81 @@
+      // so that we dont accidentally overwrite it
+      if (checkIfAgencyAddressFormIsEmpty()) {
+        fetchAgencyData();
+      }
+      else {
+        const confirmation = confirm('Do you want to overwrite the address with agency address?');
+        if (confirmation === true) {
+          fetchAgencyData();
```
formatting

Enubia commented on client/public/clientJS/wishcardEdit.js:
```diff
@@ -0,0 +1,81 @@
+  // get html nodes and input with agency details from backend
+  function updateAgencyAddressForm(agencyAddress) {
+    // get all address input elements
+    let address1 = document.getElementById('address1');
+    let address2 = document.getElementById('address2');
+    let zipCode = document.getElementById('address_zip');
+    let city = document.getElementById('address_city');
+    let country = document.getElementById('address_country');
+    let state = document.getElementById('address_state');
+    // set new values
+    address1.value = agencyAddress.address1;
+    address2.value = agencyAddress.address2;
+    zipCode.value = agencyAddress.zipcode;
+    city.value = agencyAddress.city;
+    country.value = agencyAddress.country;
+    state.value = agencyAddress.state;
+  }
+
+  function checkIfAgencyAddressFormIsEmpty() {
+    let address1 = document.getElementById('address1');
+    let address2 = document.getElementById('address2');
+    let zipCode = document.getElementById('address_zip');
+    let city = document.getElementById('address_city');
+    let country = document.getElementById('address_country');
+    let state = document.getElementById('address_state');
+    // if length is zero, input fields are empty
+    return (address1.value.length || address2.value.length || zipCode.value.length || city.value.length || country.value.length || state.value.length) === 0;
```
use jquery selectors

Enubia commented on server/routes/wishCards.js:
```diff
@@ -199,6 +199,82 @@ router.post(
+      let { childBirthday, wishItemPrice } = req.body;
+      const wishcard = await WishCardRepository.getWishCardByObjectId(req.params.id);
+
+      childBirthday = childBirthday ? new Date(childBirthday) : wishcard.childBirthday;
+      wishItemPrice = wishItemPrice ? Number(wishItemPrice) : wishcard.wishItemPrice;
+
+      await WishCardRepository.updateWishCard(wishcard._id, {
+        childBirthday,
+        wishItemPrice,
```
you can set the bithday and wishitemprice inside the parameters like you did with the address

Enubia commented on server/routes/wishCards.js:
```diff
@@ -199,6 +199,82 @@ router.post(
+    if (res.locals.user.userRole !== 'admin') {
+      const userAgency = await AgencyRepository.getAgencyByUserId(res.locals.user._id);
+      if (String(agency._id) !== String(userAgency._id)) {
+        return res.status(404).render('403');
```
why are we sending 404 and rendering 403? doesn't make sense to me

Enubia commented on server/routes/wishCards.js:
```diff
@@ -199,6 +199,82 @@ router.post(
+
+    await WishCardRepository.deleteWishCard(wishcard._id);
+
+    let url = '/wishcards/me';
+    if (res.locals.user.userRole === 'admin') {
+      url = '/wishcards/';
+    }
+
```
you can move this before the first ```res.locals.user.userRole``` and just change the logic a bit so that you assign the admin route first, than check for userRole != admin and if that's true append the ```/me``` to the route

gets rid of the second userRole check

Enubia commented on server/routes/wishCards.js:
```diff
@@ -252,6 +328,36 @@ router.get('/me', renderPermissions, async (req, res) => {
+    }
+    res.render('wishcardEdit', { wishcard, agencyAddress, childBirthday }, (error, html) => {
+      if (error) {
+        res.status(400).json({ success: false, error });
```
this renders a blank page with just json on it

render a different page here with some info for the user instead of plain json

make it pretty

Jacobjeevan commented:
- Added a 403 page
	- Changed GET routes in /wishcards/ to render 403 for errors
	- Modified tests accordingly
- Changed wishcardEdit and agencyWishCards js files to use jquery selectors because Marco likes them :+1: (half joking :laughing: )
- Added a new middleware (similar to renderPermissions) that redirects to 403 for GET routes
	- Rather than returning a JSON object

Enubia reviewed (approved):
LGTM
